### PR TITLE
Fixed regression in URIExtension.

### LINKED ISSUE #456: Regression in `UriExtensions`: `inmemory:/` URI scheme is not supported any more
After merging GH-403 into the master branch, `inmemory:/` URI scheme cannot be used for testing as those will be converted into `file:///` URI scheme.

Test case:
```java
@Test
def void testInMemory_GH_403() {
  initialize
  val uri = 'inmemory:/foo/bar.testlang';
  open(uri, '''
    type Foo { }
  ''');
  assertNotNull('''Diagnostics could be empty but never null.''', diagnostics.get(uri));
}
```

Additional details are [here](https://github.com/eclipse/xtext-core/pull/403).

### PR DESCRIPTION
**Pull request by kittaakos**
Task: #456 
Closes #456.

### CODE REVIEW

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend [hunk 1]
```diff
@@ -50,8 +254,46 @@ class UriExtensionsTest {
...
+		assertTrue(directory.exists)
+		assertTrue(directory.directory)
+		assertTrue(CharMatcher.WHITESPACE.matchesAnyOf(directory.name));
+		var uri = createEmfFileURI(directory.absolutePath).toPath.toUri
 		assertTrue(uri.isPrefix)
 	}
-}
 
+	private def createEmfURI(String uri) {
+		return org.eclipse.emf.common.util.URI.createURI(uri);
```
Comment (akosyakov) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend]: you could use static extension import

File: org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend [hunk 2]
```diff
@@ -50,8 +254,46 @@ class UriExtensionsTest {
...
+	}
+	
+	private def createEmfFileURI(String uri) {
+		return org.eclipse.emf.common.util.URI.createFileURI(uri);
```
Comment (akosyakov) [inline org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend]: the same here

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend [hunk 3]
```diff
@@ -21,32 +27,100 @@ import java.nio.charset.StandardCharsets
...
+	 * Converts the EMF URI argument into a string path.
+	 * <p>
+	 * If the argument is a URI with {@link URI#isFile() file} scheme, then the produced result value will always have this
+	 * form {@code file:///path/to/resource, otherwise it returns with the string representation
+	 * of the argument.
+	 */
 	def String toPath(URI uri) {
-		return URI.createURI(uri.path.toPath).toString
+		if (uri.file) {
+			val segments = uri.toSegments;
```
Comment (akosyakov) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend]: I am not sure about it, could it have some query params via emf URI?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend [hunk 4]
```diff
@@ -21,32 +27,100 @@ import java.nio.charset.StandardCharsets
...
+	 * </p>
+	 */
 	def URI toUri(String pathWithScheme) {
-		// URI is used to get path of the uri without scheme
-		var path = URI.createURI(pathWithScheme).path
-		return URI.createURI(path.toPath)
+		// This will take care of the encoding.
+		val uri = URI.createURI(pathWithScheme.checkNotNull('pathWithScheme'));
+		if (uri.file) {
+			val segments = uri.toSegments;
```
Comment (akosyakov) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend]: I am a bit afraid of manual parsing/serialization, we could lose some info from URIs, like query params

Comment (kittaakos) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend]: Although I totally agree with you, I am positive the current state [does not handle this either](https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend#L26).

The ball is in your court...

Comment (akosyakov) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend]: ok, the state before as well

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend [hunk 5]
```diff
@@ -21,32 +27,100 @@ import java.nio.charset.StandardCharsets
...
+			val prefix = try {
```
Comment (svenefftinge) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend]: I think the locations that assume a uri with a trailing empty segment (prefix-uri) should take care. 
Handling it in a central manner will alter URIs surprisingly in other cases, also it won't work for non existing directories but should.
We already do stuff like that here : https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext/src/org/eclipse/xtext/workspace/FileProjectConfig.xtend#L98

Comment (svenefftinge) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend]: I couldn't find a location where the `org.eclipse.xtext.util.UriUtil.isPrefixOf(URI, URI)` is used with a uri that has not been converted. Why was it added here?
If you find such place, please use `org.eclipse.xtext.util.UriUtil.toFolderURI(URI)`.

Comment (svenefftinge) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/UriExtensions.xtend]: Maybe not `isPrefixOf` but `resolve` caused the issue. We should make sure that `FileProjectConfig` is in all cases initialized with a "FolderUri" (see URIUtils).

**Comment by kittaakos** [comment]
Closing in favor of #458.

**Pull request closed without merging**